lib/backend/etcdv3/watcher.go

```diff
@@ -166,35 +112,82 @@ func (wc *watcher) watchLoop(watchClosedCh chan struct{}) {
 	wch := wc.client.etcdClient.Watch(wc.ctx, key, opts...)
 	for wres := range wch {
 		if wres.Err() != nil {
+			// A watch channel error is a terminating event, so exit the loop.
 			err := wres.Err()
 			log.WithError(err).Error("Watch channel error")
 			wc.sendError(err)
 			return
 		}
 		for _, e := range wres.Events {
+			// Convert the etcdv3 event to the equivalent Watcher event.  An error
+			// parsing the event is returned as an error.
 			if ae, err := convertWatchEvent(e, wc.list); ae != nil {
 				wc.sendEvent(ae)
 			} else if err != nil {
 				wc.sendError(err)
 			}
 		}
 	}
+	log.Debug("End of watcher.watchLoop")
+}
+
+// listCurrent retrieves the existing entries and sends an event for each listed
+func (wc *watcher) listCurrent() error {
+	log.Info("Performing initial list with no revision")
+	list, err := wc.client.List(wc.ctx, wc.list, "")
+	if err != nil {
+		return err
+	}
+
+	wc.initialRev, err = strconv.ParseInt(list.Revision, 10, 64)
+	if err != nil {
+		log.WithError(err).Error("List returned revision that could not be parsed")
+		return err
+	}
+
+	// We are sending an initial sync of entries to the watcher to provide current
+	// state.  To the perspective of the watcher, these are added entries, so set the
+	// event type to WatchAdded.
+	log.WithField("NumEntries", len(list.KVPairs)).Debug("Sending create events for each existing entry")
+	for _, kv := range list.KVPairs {
+		wc.sendEvent(&api.WatchEvent{
+			Type: api.WatchAdded,
+			New:  kv,
+		})
+	}
+	return nil
+}
+
+// terminateWatcher terminates the resources associated with the watcher.
+func (wc *watcher) terminateWatcher() {
+	log.Debug("Terminate watcher")
+	// Cancel the context - which will cancel the etcd Watch, this may have already been
+	// cancelled through an explicit Stop, but it is fine to cancel multiple times.
+	wc.cancel()
+
+	// Close the results channel.
+	close(wc.resultChan)
 
-	// The watch has been ended through a client action (e.g context cancelled or timedout).
-	//
-	// When we come to this point, it's only possible that client side ends the watch.
-	// e.g. cancel the context, close the client.etcdClient.
-	close(watchClosedCh)
+	// Set the terminated flag.
+	wc.terminated = true
 }
 
-// sendError sends an error down the errChan or waits for done notification.
+// sendError packages up the error as an event and sends it in the results channel.
 func (wc *watcher) sendError(err error) {
-	select {
-	case wc.errChan <- err:
-	case <-wc.ctx.Done():
+	// Skip cancelled errors propagating up from etcd.
+	if err == context.Canceled {
+		return
```

Comment:
I personally prefer writing `return nil` to be more explicit

Reply:
are you sure?  This method has no return type, so I don't think it is valid to return nil.

Reply:
o nvm I thought it was returning error